### PR TITLE
Remove regex usage from Identifier validation

### PR DESCRIPTION
**Pull request by wendigo**
## Description

JMH results: https://jmh.morethan.io/?gists=d2ceb76e826da7295d86efce0088a550,02f65f59e650539dad816fa36c8805f2

> Is this change a fix, improvement, new feature, refactoring, or other?

Improvement.

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

Core engine.

> How would you describe this change to a non-technical end user or system administrator?

Worse -> better (tm @skrzypo987)

## Documentation

(x) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

(x) No release notes entries required.
( ) Release notes entries required with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

**Comment by wendigo** [comment]
Screenshot 2022-03-17 at 15 13 10

**Comment by findepi** [comment]
build isn't green

**Comment by wendigo** [comment]
@findepi checkstyle. Fixed, PTAL now

File: core/trino-parser/src/main/java/io/trino/sql/tree/Identifier.java [hunk 1]
```diff
@@ -118,4 +128,17 @@ public boolean shallowEquals(Node other)
         Identifier that = (Identifier) other;
         return Objects.equals(value, that.value) && delimited == that.delimited;
     }
+
+    private static boolean isValidIdentifier(String value)
+    {
+        if (Strings.isNullOrEmpty(value)) {
+            return false;
```
Comment (findepi) [inline core/trino-parser/src/main/java/io/trino/sql/tree/Identifier.java:135]: value being null should result in NPE, not IAE

File: core/trino-parser/src/main/java/io/trino/sql/tree/Identifier.java [hunk 2]
```diff
@@ -118,4 +128,17 @@ public boolean shallowEquals(Node other)
...
+        }
+
+        if (!FIRST_CHAR_MATCHER.matches(value.charAt(0))) {
+            return false;
+        }
+
+        return TAIL_CHARS_MATCHER.matchesAllOf(value.substring(1));
```
Comment (findepi) [inline core/trino-parser/src/main/java/io/trino/sql/tree/Identifier.java:142]: nit: `.substring` is copying operation. can we do better?

Comment (wendigo) [reply core/trino-parser/src/main/java/io/trino/sql/tree/Identifier.java:142]: Of course :)

Comment (wendigo) [reply core/trino-parser/src/main/java/io/trino/sql/tree/Identifier.java]: Better now?

File: core/trino-parser/src/main/java/io/trino/sql/tree/Identifier.java [hunk 3]
```diff
@@ -131,14 +130,14 @@ public boolean shallowEquals(Node other)
...
 
-        if (!FIRST_CHAR_MATCHER.matches(value.charAt(0))) {
+        if (FIRST_CHAR_DISALLOWED_MATCHER.matches(value.charAt(0))) {
             return false;
         }
 
-        return TAIL_CHARS_MATCHER.matchesAllOf(value.substring(1));
+        // TAIL_CHARS_MATCHER is a superset of FIRST_CHAR_MATCHER
+        // so to avoid copying we are checking the first char twice
+        return TAIL_CHARS_MATCHER.matchesAllOf(value);
```
Comment (findepi) [inline core/trino-parser/src/main/java/io/trino/sql/tree/Identifier.java:141]: the "TAIL" name no longer applies

Comment (wendigo) [reply core/trino-parser/src/main/java/io/trino/sql/tree/Identifier.java:141]: As always, you are right :)

**Comment by findepi** [comment]
Added some rationale to cmt msg

**Review by findepi — APPROVED** [review]

**Comment by wendigo** [comment]
Thx @findepi !

**Pull request merged**